From: Tahsin Erdogan <tahsin@google.com>
Date: Tue, 14 Feb 2017 19:27:38 -0800
Subject: block: do not allow updates through sysfs until registration completes
Patch-Mainline: v4.11
Git-commit: b410aff2bd9f101bc096069612391aecce204cd1
References: bsc#1047027

When a new disk shows up, sysfs queue directory is created before elevator
is registered. This allows a user to attempt a scheduler switch even though
the initial registration hasn't completed yet.

In one scenario, blk_register_queue() calls elv_register_queue() and
right before cfq_registered_queue() is called, another process executes
elevator_switch() and replaces q->elevator with deadline scheduler. When
cfq_registered_queue() executes it interprets e->elevator_data as struct
cfq_data even though it is actually struct deadline_data.

Grab q->sysfs_lock in blk_register_queue() to synchronize with sysfs
callers.

[hare: ported to SLE12 SP2]

Signed-off-by: Tahsin Erdogan <tahsin@google.com>
Signed-off-by: Jens Axboe <axboe@fb.com>
Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 block/blk-sysfs.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

diff --git a/block/blk-sysfs.c b/block/blk-sysfs.c
index e08a771..de251e3 100644
--- a/block/blk-sysfs.c
+++ b/block/blk-sysfs.c
@@ -695,10 +695,13 @@ int blk_register_queue(struct gendisk *disk)
 	if (ret)
 		return ret;
 
+	/* Prevent changes through sysfs until registration is completed. */
+	mutex_lock(&q->sysfs_lock);
+
 	ret = kobject_add(&q->kobj, kobject_get(&dev->kobj), "%s", "queue");
 	if (ret < 0) {
 		blk_trace_remove_sysfs(dev);
-		return ret;
+		goto unlock;
 	}
 
 	kobject_uevent(&q->kobj, KOBJ_ADD);
@@ -707,7 +710,7 @@ int blk_register_queue(struct gendisk *disk)
 		blk_mq_register_disk(disk);
 
 	if (!q->request_fn)
-		return 0;
+		goto unlock;
 
 	ret = elv_register_queue(q);
 	if (ret) {
@@ -715,10 +718,12 @@ int blk_register_queue(struct gendisk *disk)
 		kobject_del(&q->kobj);
 		blk_trace_remove_sysfs(dev);
 		kobject_put(&dev->kobj);
-		return ret;
+		goto unlock;
 	}
-
-	return 0;
+	ret = 0;
+unlock:
+	mutex_unlock(&q->sysfs_lock);
+	return ret;
 }
 
 void blk_unregister_queue(struct gendisk *disk)
-- 
1.8.5.6

